docs/lake/design.md

```diff
@@ -0,0 +1,657 @@
+use case and workflow, we envision that some amount of overlapping segments
+would persist at small scale and always be "mixed in" with other overlapping
+segments during any key-range scan.
+
+> Note: since this style of data organization follows the LSM pattern,
+> how data is rolled up (or not) can control the degree of LSM write
+> amplification that occurs for a given workload.  There is an explicit
+> tradeoff here between overhead of merging overlapping segments on read
+> and LSM write amplification to organize the data to avoid such overlaps.
+
+> Note: we are showing here manual, CLI-driven steps to accomplish these tasks
+> but a live data pipeline would automate all of this with orchestration that
+> performs these functions via a service API, i.e., the same service API
+> used by the CLI operators.
+
+### Log
+
+Like `git log`, the command `zed lake log` prints the journal of commit
+operations.
+
+> TBD: define format of this output.  The info record context should be
+> displayed along with the adds/drops of each commit tag.
+
+## Cloud Object Naming
+
+A Zed lake semantics defined above are achieved by mapping the
+lake abstractions onto a key-value cloud obejct store.
+
+Every data element in a Zed lake is one of three fundamental object types:
+* a single-writer _immutable object_,
+* a multi-writer _commit journal_, or
+* a multi-writer _mutable object_.
+
+### Immutable Objects
+
+All data in a data pool is comprised of immutable objects, which are organized
+into data _segments_.  Each segment is composed of one or more immutable objects
+all of which share a common, globally unique identifier,
+which refer to below as `<id>`.
+
+These identifiers are [KSUIDs](https://github.com/segmentio/ksuid).
+The KSUID allocation scheme
+provides a decentralized solution for creating globally unique IDs.
+KSUIDs have embedded timestamps so the creation time of
+any object named in this way can be derived.  Also, a simple lexicographic
+sort of the KSUIDs results in a creation-time ordering (though this ordering
+is not relied on for causal relationships since clock skew can violate
+such an assumption).
+
+Segments are referred to by zero or more commits, where the commits
+are maintained in a commit journal described below.
+
+> While a zed lake is defined in terms of a cloud object store, it may also
+> be realized on top of a file system, which provides a convenient means for
+> local, small-scale deployments or test/debug workflows.  Thus, for simple use cases,
+> the complexity of running an object-store service may be avoided.
+
+#### Segments
+
+An immutable object is created by a single writer using a globally unique name
+with an embedded KSUID.  
+New objects are written in their entirety.  No updates, appends, or modifications
+may be made once an object exists.  Given these semantics, any such object may be
+trivially cached as its name or content never changes.
+
+Since the writer the name is globally unique and the
```

Comment:
I'm not sure how to best fix this, but "the writer the name" looks like mistaken wording.